File: convert_to_json.js
/**
 * Convert questions.xls (HTML table) to questions.json
 * Node.js script — no external dependencies
 */
const fs = require('fs');
const path = require('path');

const html = fs.readFileSync('questions.xls', 'utf-8');

// Column name mapping
const COL_MAP = {
  '题目类型': 'qtype',
  '选择题题干': 'stem',
  '正确答案': 'answer',
  '答案解析': 'explanation',
  '难易度': 'difficulty',
  '知识点': 'topic',
  '标签': 'tags',
  '选项数': 'opt_count',
  '选项A': 'opt_a',
  '选项B': 'opt_b',
  '选项C': 'opt_c',
  '选项D': 'opt_d',
};

// Parse HTML table rows
const rowRegex = /<tr>(.*?)<\/tr>/gs;
const cellRegex = /<t[hd][^>]*>(.*?)<\/t[hd]>/gs;

const rows = [...html.matchAll(rowRegex)].map(m => m[1]);
console.log(`Found ${rows.length} rows (including header)`);

if (rows.length < 2) {
  console.error('Not enough rows found in HTML table');
  process.exit(1);
}

// Parse header
function parseCells(rowHtml) {
  const cells = [...rowHtml.matchAll(cellRegex)].map(m => {
    // Strip all HTML tags and trim
    return m[1].replace(/<[^>]+>/g, '').trim();
  });
  return cells;
}

const headerCells = parseCells(rows[0]);
const header = headerCells.map(c => c.replace(/[\s　]/g, ''));
console.log('Header:', header);

const mapped = header.map(h => COL_MAP[h] || h);
console.log('Mapped:', mapped);

// Parse data
const data = [];
for (let i = 1; i < rows.length; i++) {
  const cells = parseCells(rows[i]);
  if (cells.length === 0) continue;

  // Skip rows with no content
  if (!cells.some(c => c.length > 0)) continue;

  const obj = {};
  for (let j = 0; j < mapped.length; j++) {
    obj[mapped[j]] = j < cells.length ? cells[j] : '';
  }

  // Validate: must have stem and valid answer (A/B/C/D)
  const stem = obj.stem || '';
  const answer = (obj.answer || '').toUpperCase().trim();
  if (stem && /^[ABCD]$/.test(answer)) {
    obj.answer = answer;
    data.push(obj);
  }
}

console.log(`Valid questions: ${data.length}`);

// Write JSON
fs.writeFileSync('questions.json', JSON.stringify(data, null, 2), 'utf-8');
console.log('Written to questions.json');

const jsonSize = JSON.stringify(data).length;
console.log(`JSON size: ${(jsonSize / 1024).toFixed(1)} KB`);

// Show sample
if (data.length > 0) {
  const q = data[0];
  console.log(`\nSample question:`);
  console.log(`  stem: ${(q.stem || '').substring(0, 80)}`);
  console.log(`  answer: ${q.answer}`);
  console.log(`  opt_a: ${(q.opt_a || '').substring(0, 40)}`);
  console.log(`  opt_b: ${(q.opt_b || '').substring(0, 40)}`);
}

File: convert_to_json.py
"""Convert questions.xls (HTML table) to questions.json"""
import re
import json

with open('questions.xls', 'r', encoding='utf-8-sig') as f:
    content = f.read()

# Extract all table rows
rows = re.findall(r'<tr>(.*?)</tr>', content, re.DOTALL)
print(f'Found {len(rows)} rows (including header)')

# Parse header from first row
header_cells = re.findall(r'<t[hd]\s+class="text">(.*?)</t[hd]>', rows[0], re.DOTALL)
if not header_cells:
    # Try simpler pattern
    header_cells = re.findall(r'<t[hd][^>]*>(.*?)</t[hd]>', rows[0], re.DOTALL)

# Clean header: strip HTML tags and whitespace
header = []
for c in header_cells:
    clean = re.sub(r'<[^>]+>', '', c).strip()
    # Remove all spaces (Chinese and English)
    clean = clean.replace(' ', '').replace('　', '')
    header.append(clean)
print(f'Header ({len(header)}): {header}')

# Map Chinese column names to internal keys
COL_MAP = {
    '题目类型': 'qtype',
    '选择题题干': 'stem',
    '正确答案': 'answer',
    '答案解析': 'explanation',
    '难易度': 'difficulty',
    '知识点': 'topic',
    '标签': 'tags',
    '选项数': 'opt_count',
    '选项A': 'opt_a',
    '选项B': 'opt_b',
    '选项C': 'opt_c',
    '选项D': 'opt_d',
}
mapped = [COL_MAP.get(h, h) for h in header]
print(f'Mapped: {mapped}')

# Parse data rows
data = []
for row in rows[1:]:
    cells = re.findall(r'<td[^>]*>(.*?)</td>', row, re.DOTALL)
    if not cells:
        continue

    vals = []
    for c in cells:
        clean = re.sub(r'<[^>]+>', '', c).strip()
        vals.append(clean)

    # Skip empty rows
    has_content = any(v for v in vals)
    if not has_content:
        continue

    obj = {}
    for i, h in enumerate(mapped):
        if i < len(vals):
            obj[h] = vals[i]
        else:
            obj[h] = ''

    # Only keep valid rows: must have stem and valid answer (A/B/C/D)
    stem = obj.get('stem', '')
    answer = obj.get('answer', '').strip().upper()
    if stem and answer in ('A', 'B', 'C', 'D'):
        obj['answer'] = answer
        data.append(obj)

print(f'Valid questions: {len(data)}')

# Write JSON
with open('questions.json', 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False)
print('Written to questions.json')

# Print sample
if data:
    print(f'\nSample question:\n  stem: {data[0].get("stem","")[:80]}\n  answer: {data[0].get("answer","")}\n  options: A={data[0].get("opt_a","")[:40]} B={data[0].get("opt_b","")[:40]}')
